File: sdk/rust/src/artifact/rsync.rs
use crate::{
    api::artifact::ArtifactSystem::{Aarch64Darwin, Aarch64Linux, X8664Darwin, X8664Linux},
    artifact::{step, Artifact, ArtifactSource},
    context::ConfigContext,
};
use anyhow::Result;
use indoc::formatdoc;

/// Build-target `rsync` file-copying tool, built from source.
#[derive(Default)]
pub struct Rsync;

impl Rsync {
    /// Creates a builder for the pinned `rsync` release.
    #[must_use]
    pub fn new() -> Self {
        Self
    }

    /// Builds the `rsync` artifact for the context's target system.
    ///
    /// # Errors
    ///
    /// Returns an error if registering the source or artifact with the build context fails.
    pub async fn build(self, context: &mut ConfigContext) -> Result<String> {
        let name = "rsync";
        let version = "3.4.1";

        let path = format!("https://sdk.vorpal.build/source/rsync-{version}.tar.gz");
        let source = ArtifactSource::new(name, &path).build();

        let step_script = formatdoc! {"
            mkdir -p \"$VORPAL_OUTPUT\"
            pushd ./source/{name}/{name}-{version}
            ./configure \
                --prefix=\"$VORPAL_OUTPUT\" \
                --disable-openssl \
                --disable-xxhash \
                --disable-zstd \
                --disable-lz4
            make
            make install",
        };

        let steps = vec![step::shell(context, vec![], vec![], step_script, vec![]).await?];
        let systems = vec![Aarch64Darwin, Aarch64Linux, X8664Darwin, X8664Linux];

        Artifact::new(name, steps, systems)
            .with_aliases(vec![format!("{name}:{version}")])
            .with_sources(vec![source])
            .build(context)
            .await
    }
}

File: sdk/typescript/src/artifact/rsync.ts
import { ArtifactSystem } from "../api/artifact/artifact.js";
import type { ConfigContext } from "../context.js";
import { Artifact, ArtifactSource } from "../artifact.js";
import { shell } from "./step.js";

/**
 * Builder for the Rsync artifact.
 *
 * Mirrors Rust `Rsync` struct in `sdk/rust/src/artifact/rsync.rs`.
 * Downloads tarball, runs configure+make with specific disable flags.
 */
export class Rsync {
  async build(context: ConfigContext): Promise<string> {
    const name = "rsync";

    const version = "3.4.1";

    const path = `https://sdk.vorpal.build/source/rsync-${version}.tar.gz`;

    const source = new ArtifactSource(name, path).build();

    const stepScript = `mkdir -p "$VORPAL_OUTPUT"
pushd ./source/${name}/${name}-${version}
./configure --prefix="$VORPAL_OUTPUT" --disable-openssl --disable-xxhash --disable-zstd --disable-lz4
make
make install`;

    const steps = [await shell(context, [], [], stepScript, [])];

    const systems = [
      ArtifactSystem.AARCH64_DARWIN,
      ArtifactSystem.AARCH64_LINUX,
      ArtifactSystem.X8664_DARWIN,
      ArtifactSystem.X8664_LINUX,
    ];

    return new Artifact(name, steps, systems)
      .withAliases([`${name}:${version}`])
      .withSources([source])
      .build(context);
  }
}
